### PR TITLE
test(e2e): Add tests for Recent Objects

### LINKED ISSUE #6150: [e2e] More automated tests for Recent Objects
#### Summary

Add automated e2e tests to cover the whole of Recent Objects functionality

### PR DESCRIPTION
**Pull request by ozyx**
Closes #6150 

### Describe your changes:

Adds tests for Recent Objects to cover the following:

- Basic smoke test
- Object renames are reflected
- Object deleted are reflected
- Clicking on object path navigates and updates url hash
- Persists on refresh
- Displays objects and aliases uniquely
- Clicking on target button scrolls object into view and highlights it
- Enforces a limit of 20 objects

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [x] Tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [ ] Changes appear to address issue?
* [ ] Reviewer has tested changes by following the provided instructions?
* [ ] Changes appear not to be breaking changes?
* [ ] Appropriate automated tests included?
* [ ] Code style and in-line documentation are appropriate?
* [ ] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [ ] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

### CODE REVIEW

File: e2e/tests/functional/recentObjects.e2e.spec.js [hunk 1]
```diff
@@ -50,14 +54,9 @@ test.describe('Recent Objects', () => {
         await page.mouse.move(0, 100);
```
Comment (unlikelyzero) [inline e2e/tests/functional/recentObjects.e2e.spec.js:54]: I can't leave a comment on line #48 but do we have a better locator for this element?

File: e2e/tests/functional/recentObjects.e2e.spec.js [hunk 2]
```diff
@@ -102,31 +101,117 @@ test.describe('Recent Objects', () => {
...
+            name: clock.name
+        });
+        await clockTreeItem.click({
+            button: 'right'
+        });
+        await page.getByRole('menuitem', {
+            name: /Create Link/
+        }).click();
+        await page.getByRole('tree', { name: 'Create Modal Tree'}).getByRole('treeitem').first().click();
+        await page.getByRole('button', {
```
Comment (unlikelyzero) [inline e2e/tests/functional/recentObjects.e2e.spec.js:174]: i think we should consider expanding the number of lines per row to allow this locator to be a one liner

File: e2e/tests/functional/recentObjects.e2e.spec.js [hunk 3]
```diff
@@ -102,31 +101,117 @@ test.describe('Recent Objects', () => {
...
+        expect(await recentObjectsList.locator('.is-alias').count()).toBe(1);
+
+        // Assert that the alias and the original's breadcrumbs are different
+        const clockBreadcrumbs = recentObjectsList.getByRole('listitem', {name: clock.name}).getByRole('navigation');
+        expect(await clockBreadcrumbs.count()).toBe(2);
+        expect(await clockBreadcrumbs.nth(0).innerText()).not.toEqual(await clockBreadcrumbs.nth(1).innerText());
+    });
+    test("Enforces a limit of 20 recent objects", async ({ page }) => {
+        // Creating 21 objects takes a while, so increase the timeout
+        test.slow();
```
Comment (unlikelyzero) [inline e2e/tests/functional/recentObjects.e2e.spec.js:196]: this decorator used to work at one point so you could throw it onto the test itself. i.e.

```test.slow("Enforces a limit```

File: e2e/tests/functional/recentObjects.e2e.spec.js (hunk 2 above)
Comment (ozyx) [reply e2e/tests/functional/recentObjects.e2e.spec.js:174]: We can compress this to a single line, that was me manually adding endlines.

Top of build 8 we should get prettier implemented. I have it mostly done on a branch

File: e2e/tests/functional/recentObjects.e2e.spec.js [hunk 4]
```diff
@@ -102,31 +101,117 @@ test.describe('Recent Objects', () => {
...
+
+        // Create 19 more objects (2 in beforeEach() + 19 new = 21 total)
```
Comment (unlikelyzero) [inline e2e/tests/functional/recentObjects.e2e.spec.js:198]: If this is the case, i would add a        
```
 // Assert that two recent objects are already available
  expect(await recentObjectsList.getByRole('listitem', { name: clock.name }).count()).toBe(2);
```

File: src/ui/layout/RecentObjectsListItem.vue [hunk 5]
```diff
@@ -54,6 +54,7 @@
     <div class="c-recentobjects-listitem__target-button">
         <button
             class="c-icon-button icon-target"
+            :aria-label="`Open and scroll to ${domainObject.name}`"
```
Comment (unlikelyzero) [inline src/ui/layout/RecentObjectsListItem.vue:57]: back to our `id` conversation, is it possible to reference domainObject.name as an id and use aria-labelledby to compute this aria-label?

File: e2e/tests/functional/recentObjects.e2e.spec.js [hunk 6]
```diff
@@ -102,31 +101,117 @@ test.describe('Recent Objects', () => {
...
+
+        // Wait for highlight animation to end
+        await waitForAnimations(clockTreeItem.locator('.c-tree__item'));
+
+        // Assert that the Clock treeitem is no longer highlighted
+        await expect(clockTreeItem.locator('.c-tree__item')).not.toHaveClass(/is-targeted-item/);
     });
-    test.fixme("Tests for context menu actions from recent objects", async ({ page }) => {
+    test("Persists on refresh", async ({ page }) => {
+        assertInitialRecentObjectsListState();
```
Comment (unlikelyzero) [inline e2e/tests/functional/recentObjects.e2e.spec.js:150]: ha!

File: src/ui/components/ObjectPath.vue [hunk 7]
```diff
@@ -24,7 +24,7 @@
 <ul
     v-if="orderedPath.length"
     class="c-location"
-    :aria-label="`${domainObject.name} Breadcrumb`"
+    :aria-label="`${domainObject.name}`"
```
Comment (unlikelyzero) [inline src/ui/components/ObjectPath.vue:27]: oh nice

File: e2e/tests/functional/recentObjects.e2e.spec.js [hunk 8]
```diff
@@ -102,31 +101,117 @@ test.describe('Recent Objects', () => {
...
+
+        // Create 19 more objects (2 in beforeEach() + 19 new = 21 total)
+        for (let i = 0; i < 19; i++) {
+            await createDomainObjectWithDefaults(page, {
+                type: "Clock"
+            });
+        }
+
+        // Assert that the list contains 20 objects
+        expect(await recentObjectsList.locator('.c-recentobjects-listitem').count()).toBe(20);
```
Comment (unlikelyzero) [inline e2e/tests/functional/recentObjects.e2e.spec.js:206]: is there anything else worth asserting at this point given the test execution time?

File: e2e/tests/functional/recentObjects.e2e.spec.js [hunk 9]
```diff
@@ -102,31 +101,117 @@ test.describe('Recent Objects', () => {
...
-        expect(page.url()).toMatch(new RegExp(`.*mine\?.*`));
+        expect(page.url()).toMatch(new RegExp(`.*mine?.*`));
     });
-    test.fixme("Clicking on the 'target button' scrolls the object into view in the tree and highlights it", async ({ page }) => {
+    test("Clicking on the 'target button' scrolls the object into view in the tree and highlights it", async ({ page }) => {
+        const clockTreeItem = page.getByRole('tree', { name: 'Main Tree'}).getByRole('treeitem', { name: clock.name });
+        const folderTreeItem = page.getByRole('tree', { name: 'Main Tree'})
+            .getByRole('treeitem', {
+                name: folderA.name,
+                expanded: true
```
Comment (unlikelyzero) [inline e2e/tests/functional/recentObjects.e2e.spec.js:130]: this is really cool

**Review by unlikelyzero — COMMENTED** [review]
This is really comprehensive. I just wish we had the codecov inline reporter to show us what was missed.

Just a few questions, and a suggestion for a better locator on line 48.

I think we might want to add a single visual test for a "deeply nested object" within the tree and recent objects components. Could you put in a stub?

File: e2e/tests/functional/recentObjects.e2e.spec.js (hunk 3 above)
Comment (ozyx) [reply e2e/tests/functional/recentObjects.e2e.spec.js:196]: [Not anymore, apparently.](https://playwright.dev/docs/api/class-test#test-slow-1)

File: src/ui/layout/RecentObjectsListItem.vue (hunk 5 above)
Comment (ozyx) [reply src/ui/layout/RecentObjectsListItem.vue:57]: `domainObject.name` is actually not unique, so wouldn't make a very good id, but I wonder if we can use the `navigationPath` instead as an id and reference that? we could then use aria-labelledby.

lemme play around with that

Comment (ozyx) [reply src/ui/layout/RecentObjectsListItem.vue:57]: So we can use something like `recent-object:${navigationPath}` to make a unique id for these list items, but we would still be missing a meaningful, unique a11y label to label it with. When you get some time let's chat about and revisit this

File: e2e/tests/functional/recentObjects.e2e.spec.js (hunk 1 above)
Comment (ozyx) [reply e2e/tests/functional/recentObjects.e2e.spec.js:56]: updated to not use `nth-child`. A bit more deterministic

File: e2e/tests/functional/recentObjects.e2e.spec.js (hunk 8 above)
Comment (ozyx) [reply e2e/tests/functional/recentObjects.e2e.spec.js:216]: Updated the test to do another test for the "target" button, this time on an extremely deeply nested object

**Review by unlikelyzero — APPROVED** [review]

**Pull request merged**